### ISSUE TITLE
Step method of Number

### ISSUE THREAD

**Issue by gmarcais** [crystal-lang/crystal#6709]
As briefly mentioned in #6669, the `step` method from `Number` has the same issues as `upto` and `downto` methods used to have (possibility of infinite loop because of arithmetic overflow).

In addition, the behavior regarding types is not well specified. The receiving object, the `to:` and `by:` arguments can all have different types, and it is difficult to know, for example, what the proper behavior should be when the argument `by` is a floating point but the receiving object is an integer.

The Ruby documentation for the method `Numeric#step` says:
> If all the arguments are integers, the loop operates using an integer counter.
> 
> If any of the arguments are floating point numbers, all are converted to floats, and the loop is executed floor(n + n*Float::EPSILON) + 1 times, where n = (limit - num)/step.
> 
> Otherwise, the loop starts at num, uses either the less-than (<) or greater-than (>) operator to compare the counter against limit, and increments itself using the + operator.

In the context of Crystal, if we take Ruby as a guide, maybe there should be multiple `step` methods: one in `Number` with the default behavior, overloaded in the `Int` and `Float` classes.

Also, the step function of Ruby does not have the infinite loop behavior with integers as the integers do not overflow. But if one were to implement fixed size integers, the default implementation is also buggy:
```Ruby
# Define a byte in Ruby and minimal behavior
class UInt8 < Numeric
  attr_reader :x
  def initialize(x); @x = x & 0xff; end
  def >(y); @x > y.x; end
  def +(y);
    case y
    when UInt8
      UInt8.new(@x + y.x)
    else
      UInt8.new(@x + y)
    end
  end
end

UInt8.new(253).step(to: UInt8.new(255)) { |i| p i } # Infinite loop
```

This may be hard to avoid as the system does not know how the number like class behaves (like a fixed int, like a big int, like a floating point, signed / unsigned, etc.). This would be an argument to not have a version of `step` in `Number`, and only the versions in `Int` and `Float`.

**Comment by ttowncompiled** [comment]
I was actually going to post a new issue about Ranges, but, if I may, I think this is a good post to extend from. I noticed that in Range, is uses the `.succ` method to implement the `.step` method. There are many classes who have implemented `.succ` including Int, String, Char, etc. However, not Float which creates an issue if we use the call `.step(1.5)`. I understand the reason for wanting to allow Ranges which have mixes of Ints, Floats, and Strings (within reason of course). However, this prevents Crystal from being able to require that the `Range (B,E)` can logically iterate from `B` to `E`. And it also introduces ambiguity in the use of Strings as not all Strings have a successor, only certain subsets of strings such as alphabets. I understand that this was also implemented in Ruby, but I'll get to that in a second.

With that said, I like the use of `.succ`. However, it would seem that there isn't an abstract class or interface which requires the `.succ` method. My suggestion is to implement such a class which has the `.succ` and `.pred` methods. The Range and other implementations of `.step`, `.downto`, `.upto` could then be typed to accept classes which provide this type. This would also allow developers to define new types of Ranges. Ranges that can range over their own custom types so long as they implement a SuccIterable (or whatever it will be called) for their custom type. This would provide further extensibility to Crystal. It would also allow there to be restriction on which Strings can be combined together in a Range so that Alpha Strings can be combined, while restricting the use of Alpha strings with non-Alpha Strings such as `("9".."z")` which currently falls into an infinite loop. This would also restrict mixes of Numbers and Strings while allowing mixes of Numbers such as Ints and Floats.

In addition, I was going to suggest further adding `.succ(by: Number)` and `.pred(by: Number)` methods to this class. The call `.succ(2)` could be equivalent to two calls of `.succ`. However, it would also allow for classes such as Float and Int to have more flexibility because a Float could then have a `.succ` method which would assume `.succ(by=1)`, but would allow any Float to be passed as the `by` arg, and Int could accept `.succ(1.5)` since it could then compute `self+1.5` as the successor (converting itself to a Float of course). This would allow Ranges to accept Float inputs to step which they currently do not.

https://play.crystal-lang.org/#/r/50ad

It would allow a Float to be passed to a `.step` call because the implementation of the `.step` method could use the `.succ(by)` method instead of combining `.times` with `.succ`.

WDYT?

Warmest regards.

**Comment by straight-shoota** [comment]
As far as I can see, most of the problems mentioned here are either already fixed (arithmetic overflow raises) or addressed in #10130.

The remaining parts are mostly covered in #9339, particularly the proposal in https://github.com/crystal-lang/crystal/issues/9339#issuecomment-749281784 should be an adequate alternative solution. Standardizing on `#step(*to, by)` is more powerful than `#succ(by)`, `#pred(by)`.

A nice aspect: types that don't implement `+` and `-` methods can use a generic `#step` implementation based on `#succ` and `#pred` like this:
```cr
def step(*, to limit = nil, by step, &) : Nil
  current = self
  direction = by.sign
  by = by.abs
  while limit.nil? || ((limit <=> current).try(&.sign) == direction)
    yield current
    if direction == 1
	  by.times { current = current.succ }
    else
      by.times { current = current.pred }
    end
  end
  yield current if current == limit
end
```
(The code is untested, insufficient and probably doesn't work as it. It's just for illustration.)
